L1Trigger/Phase2L1ParticleFlow/src/corrector.cc

```diff
@@ -0,0 +1,193 @@
+        break;  // no EMF bins beyond eta = 3
+      }
+    }
+  }
+  timer.stop();
+  if (debug)
+    std::cout << "Read " << ngraphs << " graphs from " << lFile->GetPath() << " in " << timer.realTime() << " s"
+              << std::endl;
+}
+
+l1tpf::corrector::corrector(const TH1 *index, float emfMax)
+    : index_((TH1 *)index->Clone("INDEX")),
+      is2d_(index->InheritsFrom("TH2")),
+      neta_(index->GetNbinsX()),
+      nemf_(is2d_ ? index->GetNbinsY() : 1),
+      emfMax_(emfMax) {
+  index_->SetDirectory(nullptr);
+  corrections_.resize(neta_ * nemf_);
+  std::fill(corrections_.begin(), corrections_.end(), nullptr);
+}
+
+l1tpf::corrector::~corrector() {
+  for (TGraph *&p : corrections_) {
+    delete p;
+    p = nullptr;
+  }
+  corrections_.clear();
+}
+
+l1tpf::corrector::corrector(corrector &&corr)
+    : index_(std::move(corr.index_)),
+      corrections_(std::move(corr.corrections_)),
+      is2d_(corr.is2d_),
+      neta_(corr.neta_),
+      nemf_(corr.nemf_),
+      emfMax_(corr.emfMax_) {}
+
+l1tpf::corrector &l1tpf::corrector::operator=(corrector &&corr) {
+  std::swap(is2d_, corr.is2d_);
+  std::swap(neta_, corr.neta_);
+  std::swap(nemf_, corr.nemf_);
+  std::swap(emfMax_, corr.emfMax_);
+
+  index_.swap(corr.index_);
+  corrections_.swap(corr.corrections_);
+  return *this;
+}
+
+float l1tpf::corrector::correctedPt(float pt, float emPt, float eta) const {
+  float total = std::max(pt, emPt), abseta = std::abs(eta);
+  float emf = emPt / total;
+  if (emfMax_ > 0 && emf > emfMax_)
+    return total;  // no correction
+  unsigned int ieta = std::min(std::max<unsigned>(1, index_->GetXaxis()->FindBin(abseta)), neta_) - 1;
+  unsigned int iemf =
+      is2d_ && abseta < 3.0 ? std::min(std::max<unsigned>(1, index_->GetYaxis()->FindBin(emf)), nemf_) - 1 : 0;
+  TGraph *graph = corrections_[ieta * nemf_ + iemf];
+  if (!graph) {
+    throw cms::Exception("RuntimeError") << "Error trying to read calibration for eta " << eta << " emf " << emf
+                                         << " which are not available." << std::endl;
+  }
+  float ptcorr = std::min<float>(graph->Eval(total), 4 * total);
+  return ptcorr;
+}
+
+void l1tpf::corrector::correctPt(l1t::PFCluster &cluster, float preserveEmEt) const {
+  float newpt = correctedPt(cluster.pt(), cluster.emEt(), cluster.eta());
+  cluster.calibratePt(newpt, preserveEmEt);
+}
+
+void l1tpf::corrector::setGraph(const TGraph &graph, int ieta, int iemf) {
+  char buff[32];
+  if (is2d_) {
+    snprintf(buff, 31, "eta_bin%d_emf_bin%d", (unsigned int)(ieta + 1), (unsigned int)(iemf + 1));
+  } else {
+    snprintf(buff, 31, "eta_bin%d", (unsigned int)(ieta + 1));
+  }
+  TGraph *gclone = (TGraph *)graph.Clone(buff);
+  delete corrections_[ieta * nemf_ + iemf];
+  corrections_[ieta * nemf_ + iemf] = gclone;
+}
+
+void l1tpf::corrector::writeToFile(const std::string &filename, const std::string &directory) const {
+  TFile *lFile = TFile::Open(filename.c_str(), "RECREATE");
```

Comment:
potentially not thread safe

Reply:
the interface to write to files is not used when running CMSSW